Preserve seek time and other parameters to `amp-youtube` from YouTube embeds

dhaval-parekh opened the pull request:
## Summary

<!-- Please reference the issue(s) this PR fixes, if one exists. For significant changes, opening an issue first for discussion is usually preferable. -->
Fixes #4518 

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

dhaval-parekh committed:
Transfer seek time to AMP youtube component from youtube url

dhaval-parekh committed:
Add php unit test cases

westonruter commented:
There are additional parameters which the embed API supports: https://developers.google.com/youtube/player_parameters

For example, if you add a “raw embed” via the Custom HTML block with:

```html
<iframe width="560" height="315" src="https://www.youtube.com/embed/0zM3nApSvMg?controls=0&amp;autoplay=1" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```

This should result in an `amp-youtube` element being added with `data-param-controls="0"` and `data-param-autoplay="1"`.

I think any query parameters on the YouTube embed URL should be copied to `data-pram-*` attributes: https://amp.dev/documentation/components/amp-youtube/#data-param-*

For the YouTube Embed block specifically, it appears that only the `start` parameter is passed along in oEmbed requests. So this is only relevant to YouTube raw embeds. 

Humm, but actually in looking at `AMP_YouTube_Embed_Handler` it doesn't actually implement the `sanitize_raw_embeds` method so these `iframe` video embeds are being passed through as `amp-iframe` elements instead of `amp-youtube`. So they work right now, but I think it would be better if we implemented `sanitize_raw_embeds` to convert these into `amp-youtube`.

westonruter commented:
cc @pierlon for his thoughts on this since he's done a lot of work with embeds in the past

dhaval-parekh committed:
Convert youtube iframe into amp-youtube

dhaval-parekh committed:
Add additional data params for amp-youtube components

dhaval-parekh committed:
Add unit test case for youtube embeds

dhaval-parekh committed:
Fix the issues from unit test cases

dhaval-parekh committed:
Fix the issues from unit test cases

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -77,6 +79,76 @@ public function unregister_embed() {
 	}
 
+	/**
+	 * Sanitize HTML that are not added via Gutenberg.
```
Any embedded content added via Gutenberg can be processed in this method as well since it will be called after the post content has been rendered.

As I work on each embed I try to get rid of the embed filters and refactor the `::sanitize_raw_embeds()` method to handle those filtered embeds. This is because for some embeds, WordPress wraps them in `<p>` tags, which makes them inconsistent with their non-AMP embed variant and in certain cases makes the AMP component not render correctly on the page (see https://github.com/ampproject/amp-wp/issues/4358 and https://github.com/ampproject/amp-wp/issues/4450, for example).

I've started https://github.com/ampproject/amp-wp/pull/4650 to resolve that issue, but that PR is quite outdated now and needs some updates before it can be ready for review again. In that PR you can see that I did refactor this class to remove the embed filters and handle sanitization in the [`::sanitize_raw_embed()` method](https://github.com/ampproject/amp-wp/blob/0860f489b9e59357abb1c2941d13030b194b12be/includes/embeds/class-amp-youtube-embed-handler.php#L90-L124), so that should be a guide for you.

dhaval-parekh replied:
Hey @pierlon 

I have tried removing `filter_embed_oembed_html()`, But after that, during the `the_content` filter it removes extra query parameters from the youtube URL. Like
From 
```
http://www.youtube.com/watch?v=kfVsfOSbJY0&hl=en&fs=1&w=425&h=349
```
To 
```
<p><iframe title="Rebecca Black - Friday" width="980" height="551" src="https://www.youtube.com/embed/kfVsfOSbJY0?feature=oembed" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe></p>
```

Because of that additional param will be removed and won't be able to set in `<amp-youtube/>` component during the `::sanitize_raw_embeds()`

So, I think `filter_embed_oembed_html()` is necessary.

Let me know your thoughts on this.

pierlon replied:
Ah yes you're right, I overlooked that.

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -77,6 +79,76 @@ public function unregister_embed() {
+	 */
+	public function sanitize_raw_embeds( Document $dom ) {
+
+		$nodes = $dom->xpath->query( '//iframe[ contains( @src, "youtu" ) ]' );
```
This XPath query could be made more strict. Currently this will match any `iframe[src]` that contains the string `youtu`, which means it would match URLs such as "https://mysite.com/blog/youtube-iz-da-best".

The iframe `src` used in the YouTube oEmbed ([for example](https://www.youtube.com/oembed?url=https://www.youtube.com/watch?v=q4xKvHANqjk)) begins with `https://www.youtube.com` URL, so we could ensure we only match that by doing something like:

```suggestion
		$nodes = $dom->xpath->query( '//iframe[ starts-with( @src, "https://www.youtube.com" ) ]' );
```

dhaval-parekh replied:
Yes, the current XPath query will fetch all iframe with URL that contains "youtu" including the example that you gave.
But it won't be converted. Since, we also do domain checks before it.

As you mention, the code example that you gave will only fetch an iframe that begins with `https://www.youtube.com`.  that will not work with different versions of the youtube URL in an iframe.

pierlon replied:
> But it won't be converted. Since, we also do domain checks before it.

Right, but rather than iterating through potential unwanted elements and checking to see if the `src` is correct, we could do that upfront with the XPath query.

> As you mention, the code example that you gave will only fetch an iframe that begins with https://www.youtube.com. that will not work with different versions of the youtube URL in an iframe.

Yes you're right, those variations can be added to the XPath query as well. I did something similar to that [here](https://github.com/ampproject/amp-wp/blob/0860f489b9e59357abb1c2941d13030b194b12be/includes/embeds/class-amp-youtube-embed-handler.php#L75-L82) but essentially it would be something like:


```php
$applicable_domains = [ 'youtu.be', 'youtube.com', 'youtube-nocookie.com' ];

$query_segments = array_map(
	static function ( $domain ) {
		return sprintf( 'starts-with( @src, "https://www.%1$s/" ) or starts-with( @src, "https://%1$s/" )', $domain );
	},
	$applicable_domains
);
$query          = implode( ' or ', $query_segments );

$nodes = $dom->xpath->query( sprintf( '//iframe[ %s ]', $query ) );
```

That should essentially get rid of having to do the domain check as it's now being done upfront.

dhaval-parekh replied:
Thanks, @pierlon for the feedback. I have updated the PR. Please have a look.

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -77,6 +79,76 @@ public function unregister_embed() {
+
+		$nodes = $dom->xpath->query( '//iframe[ contains( @src, "youtu" ) ]' );
+
+		foreach ( $nodes as $node ) {
```
This gives the IDE more context about the variable since it doesn't detect that it's a DOM element we're working with.

```suggestion
		/** @var Element $node */
		foreach ( $nodes as $node ) {
```

(You'd need to import `\AmpProject\Dom\Element`)

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -77,6 +79,76 @@ public function unregister_embed() {
+			);
+		}
+
+		$this->did_convert_elements = true;
```
AFAIK this value is not being read anywhere and so can be removed.

```suggestion
```

cc @westonruter

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -151,10 +230,74 @@ public function render( $args, $url ) {
+			$attributes['title'] = $args['title'];
+		}
+
+		$allowed_data_params = [
+			'cc_lang_pref',
+			'cc_load_policy',
+			'color',
+			'controls',
+			'disablekb',
+			'enablejsapi',
+			'end',
+			'fs',
+			'hl',
+			'iv_load_policy',
+			'list',
+			'listType',
+			'modestbranding',
+			'origin',
+			'playlist',
+			'playsinline',
+			'rel',
+			'widget_referrer',
+		];
```
I'm not sure if we want to hardcode all the allowed data params here since we don't control whenever they become deprecated or new ones are added. Instead, we could allow all `data-param-*` to pass through but sanitize their values.

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -151,10 +230,74 @@ public function render( $args, $url ) {
+			}
+		}
+
+		if ( ! empty( $args['start'] ) && 0 < intval( $args['start'] ) ) {
+			$attributes['data-param-start'] = intval( $args['start'] );
```
It's recommended to use type conversion here, as it's much faster than using `intval()`.

```suggestion
		if ( ! empty( $args['start'] ) && 0 < (int)$args['start'] ) {
			$attributes['data-param-start'] = (int)$args['start'];
```

pierlon reviewed (changes requested):
Thanks for working on this @dhaval-parekh.

Left a few comments here that should help move this PR forward.

dhaval-parekh committed:
Use type casting instead of intval()

dhaval-parekh committed:
Add inline comments, remove unwanted code

dhaval-parekh committed:
Update data-param for amp-youtube

dhaval-parekh committed:
Update XPath query to fetch youtube iframe

dhaval-parekh committed:
Fix php unit test cases

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -77,6 +80,90 @@ public function unregister_embed() {
 	}
 
+	/**
+	 * Sanitize HTML that are not added via Gutenberg.
```
This is somewhat incorrect as HTML produced by Gutenberg could also be handled by this method if WordPress didn't convert the URL into an embed, for example.

```suggestion
	 * Sanitize YouTube raw embeds.
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -77,6 +80,90 @@ public function unregister_embed() {
+	}
+
+	/**
+	 * To AMP youtube component from DOM Document.
```
```suggestion
	 * Parse embed attributes and return an AMP YouTube component.
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -77,6 +80,90 @@ public function unregister_embed() {
+	 *
+	 * @param Document $dom  Document DOM.
+	 * @param string   $html HTML markup of youtube iframe.
+	 * @param string   $url  Youtube URL.
```
```suggestion
	 * @param string   $html HTML markup of the YouTube iframe.
	 * @param string   $url  YouTube URL.
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -77,6 +80,90 @@ public function unregister_embed() {
+	 * @param string   $html HTML markup of youtube iframe.
+	 * @param string   $url  Youtube URL.
+	 *
+	 * @return DOMElement|false DOMElement on success, Otherwise false.
```
```suggestion
	 * @return DOMElement|false AMP component, otherwise `false`.
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -77,6 +80,90 @@ public function unregister_embed() {
+	 */
+	public function sanitize_raw_embeds( Document $dom ) {
+
+		$applicable_domains = [ 'youtu.be', 'youtube.com', 'youtube-nocookie.com' ];
```
This array of domains is also repeated in the `get_video_id_from_url()` method, so you could extract it into a class property and replace both instances with it.

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -151,10 +245,59 @@ public function render( $args, $url ) {
+		$query_vars  = [];
+		$query_param = wp_parse_url( $url, PHP_URL_QUERY );
+		wp_parse_str( $query_param, $query_vars );
+		$query_vars = ( ! empty( $query_vars ) && is_array( $query_vars ) ) ? $query_vars : [];
```
An `is_array()` check is sufficient, no?

```suggestion
		$query_vars = is_array( $query_vars ) ? $query_vars : [];
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -151,10 +245,59 @@ public function render( $args, $url ) {
+		wp_parse_str( $query_param, $query_vars );
+		$query_vars = ( ! empty( $query_vars ) && is_array( $query_vars ) ) ? $query_vars : [];
+
+		$excluded_param = [ 'autoplay', 'loop', 'start', 'v', 'vi', 'w', 'h' ];
+
+		foreach ( $query_vars as $key => $value ) {
+
+			if ( in_array( $key, $excluded_param, true ) ) {
+				continue;
+			}
+
+			$attributes[ "data-param-$key" ] = sanitize_text_field( $value );
+		}
+
+		foreach ( [ 'autoplay', 'loop' ] as $param ) {
+			if ( isset( $query_vars[ $param ] ) ) {
+				$attributes[ $param ] = $query_vars[ $param ];
+			}
+		}
+
+		if ( ! empty( $args['start'] ) && 0 < (int) $args['start'] ) {
+			$attributes['data-param-start'] = (int) $args['start'];
+		}
```
These could be combined into one for loop:

```suggestion
		$excluded_params = [ 'v', 'vi', 'w', 'h' ];

		foreach ( $query_vars as $key => $value ) {
			if ( in_array( $key, $excluded_params, true ) ) {
				continue;
			}

			if ( in_array( $key, [ 'autoplay', 'loop' ], true ) ) {
				$attributes[ $key ] = $value;
				continue;
			}

			if ( 'start' === $key && 0 < (int) $args['start'] ) {
				$attributes['data-param-start'] = (int) $args['start'];
				continue;
			}

			$attributes[ "data-param-$key" ] = esc_attr( $value );
		}
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -151,10 +245,59 @@ public function render( $args, $url ) {
 		);
 
-		if ( empty( $args['video_id'] ) ) {
+		$attributes = array_merge(
+			[ 'data-videoid' => $args['video_id'] ],
+			wp_array_slice_assoc( $args, [ 'layout', 'width', 'height' ] )
+		);
```
What about directly creating a new array instead of merging and slicing?

```suggestion
		$attributes = [
			'data-videoid' => $args['video_id'],
			'layout'       => $args['layout'],
			'width'        => $args['width'],
			'height'       => $args['height'],
		];
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -77,6 +80,90 @@ public function unregister_embed() {
+			);
+		}
+
+		return AMP_DOM_Utils::create_node(
+			$dom,
+			'amp-youtube',
+			$attributes
+		);
```
Don't forget the placeholder element:

```suggestion
		$placeholder_node = AMP_DOM_Utils::create_node(
			$dom,
			Tag::A,
			[
				Attribute::PLACEHOLDER => '',
				Attribute::HREF => esc_url_raw( $url ),
			]
		);

		$amp_component = AMP_DOM_Utils::create_node(
			$dom,
			'amp-youtube',
			$attributes
		);
		$amp_component->appendChild( $placeholder_node );

		return $amp_component;
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -167,13 +310,7 @@ public function render( $args, $url ) {
-		if ( ! empty( $args['title'] ) ) {
-			$attributes['title'] = $args['title'];
-		}
+		$args['placeholder'] = ( ! empty( $args['placeholder'] ) ) ? $args['placeholder'] : '';
```
The `placeholder` argument would always be set by `parse_props()`, no? Or perhaps it would be better to set the `placeholder` argument here.

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -236,6 +373,50 @@ private function get_video_id_from_url( $url ) {
 	}
 
+	/**
+	 * To get start time of youtube video in second.
+	 *
+	 * @param string $url Youtube URL.
+	 *
+	 * @return int Start time in second.
```
```suggestion
	 * Get the start time of the YouTube video in seconds.
	 *
	 * @param string $url YouTube URL.
	 *
	 * @return int Start time in seconds.
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -236,6 +373,50 @@ private function get_video_id_from_url( $url ) {
+		$parsed_url = wp_parse_url( $url );
+
+		if ( ! empty( $parsed_url['query'] ) ) {
+			wp_parse_str( $parsed_url['query'], $query_vars );
```
```suggestion
			$query_vars = [];
			wp_parse_str( $parsed_url['query'], $query_vars );
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -236,6 +373,50 @@ private function get_video_id_from_url( $url ) {
+			wp_parse_str( $parsed_url['query'], $query_vars );
+
+			if ( ! empty( $query_vars['start'] ) && 0 < (int) $query_vars['start'] ) {
+				$start_time = (int) $query_vars['start'];
```
Why not return from here?

```suggestion
				return (int) $query_vars['start'];
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -236,6 +373,50 @@ private function get_video_id_from_url( $url ) {
+			}
+		}
+
+		if ( empty( $start_time ) && ! empty( $parsed_url['fragment'] ) ) {
```
`$start_time` would always be `0` (thus "empty") here, no?

```suggestion
		if ( ! empty( $parsed_url['fragment'] ) ) {
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -236,6 +373,50 @@ private function get_video_id_from_url( $url ) {
+		}
+
+		if ( empty( $start_time ) && ! empty( $parsed_url['fragment'] ) ) {
+			$regex = '/^t=(?<minutes>[0-9])+m(?<seconds>[0-9]+)s$/iU';
```
```suggestion
			$regex = '/^t=(?<minutes>\d+)m(?<seconds>\d+)s$/';
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -236,6 +373,50 @@ private function get_video_id_from_url( $url ) {
+
+			preg_match( $regex, $parsed_url['fragment'], $matches );
+
+			if ( ! empty( $matches ) ) {
```
Or alternatively (to make sure the match groups do indeed exist):

```suggestion
			if ( isset( $matches['minutes'], $matches['seconds'] ) ) {
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -132,17 +219,24 @@ private function parse_props( $html, $url, $video_id ) {
-	 * @param string $url  URL.
-	 * @return string Rendered.
+	 * @param array  $args amp-youtube component arguments.
+	 * @param string $url  Youtube URL.
```
```suggestion
	 * @param string $url  YouTube URL.
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -236,6 +373,50 @@ private function get_video_id_from_url( $url ) {
+	/**
+	 * To get start time of youtube video in second.
+	 *
+	 * @param string $url Youtube URL.
```
```suggestion
	 * @param string $url YouTube URL.
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -77,6 +80,90 @@ public function unregister_embed() {
+	 */
+	public function process_embed( Document $dom, $html, $url ) {
+
+		$id = $this->get_video_id_from_url( $url );
```
I just realized the AMP component can also [accept a YouTube channel id](Youtube channel id), so we need to provide support for that as well. So if there is no video ID, then check for a live channel ID.

From the docs:

> For example, in this URL: `https://www.youtube.com/embed/live_stream?channel=UCB8Kb4pxYzsDsHxzBfnid4Q`, `UCB8Kb4pxYzsDsHxzBfnid4Q` is the channel id. You can provide a `data-live-channelid` instead of a `data-videoid` attribute to embed a stable url for a live stream instead of a video.

dhaval-parekh replied:
@pierlon I have added support for live streaming channels for iframe added as custom HTML markup.
However, WordPress by default does not provide support for embedding YouTube URLs (It will just render as text)

pierlon replied:
If WordPress never supported it I think it's safe to say we don't have to either.

> @pierlon I have added support for live streaming channels for iframe added as custom HTML markup.

I tried this but it parsed an incorrect video ID. Eg:

```html
<iframe src="https://www.youtube.com/embed/live_stream?channel=12345" allowfullscreen="" width="560" height="315"></iframe>
```

Gives:

```html
<amp-youtube data-videoid="live_stream" layout="responsive" width="560" height="315" data-param-channel="12345" ...</amp-youtube>
```

Instead, the `data-videoid` attribute should not have been set, nor the `data-param-channel` attribute. Perhaps we could add a condition to say if the parsed video ID is `live_stream` then it's invalid and bail.

dhaval-parekh replied:
Have addressed this issue.

pierlon reviewed (changes requested):
Looking good so far! Just a few comments and suggestions to be made.

dhaval-parekh committed:
Apply suggestions from PR

dhaval-parekh committed:
Update amp youtibe embed handler

dhaval-parekh committed:
update unit test cases

dhaval-parekh committed:
Update and fix unit test cases

dhaval-parekh committed:
Remove redundant code

dhaval-parekh committed:
Add support of live channel when added as raw iframe

dhaval-parekh committed:
Fix unit test cases

dhaval-parekh committed:
Add unit test cases for live streaming channel

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -82,117 +94,313 @@ public function unregister_embed() {
-			return $cache;
+
+		if ( empty( $cache ) || empty( $url ) ) {
+			return '';
```
Let's return the `$cache` so that we don't unintentionally break any other plugins using this filter.

```suggestion
			return $cache;
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -82,117 +94,313 @@ public function unregister_embed() {
-			return null;
+	public function render( $html, $url ) {
+
+		$attributes = $this->prepare_attributes( $url );
```
All methods that call the `render()` and `prepare_attributes()` methods have already determined the video ID, so I'd suggest passing it to them so as not to re-parse it from the URL each time. Something like:

```diff
--- a/includes/embeds/class-amp-youtube-embed-handler.php	(revision db1bd35807612e6a89beba31fa7d05a2be5c049d)
+++ b/includes/embeds/class-amp-youtube-embed-handler.php	(date 1625728551158)
@@ -109,20 +109,21 @@
 			return $cache;
 		}
 
-		return $this->render( $cache, $url );
+		return $this->render( $cache, $url, $video_id );
 	}
 
 	/**
 	 * Convert YouTube iframe into AMP YouTube component.
 	 *
-	 * @param string $html HTML markup of YouTube iframe.
-	 * @param string $url  YouTube URL.
+	 * @param string $html      HTML markup of YouTube iframe.
+	 * @param string $url       YouTube URL.
+	 * @param string $video_id  YouTube video ID.
 	 *
 	 * @return string HTML markup of AMP YouTube component.
 	 */
-	public function render( $html, $url ) {
+	public function render( $html, $url, $video_id ) {
 
-		$attributes = $this->prepare_attributes( $url );
+		$attributes = $this->prepare_attributes( $url, $video_id );
 
 		$iframe_props = [ 'title', 'height', 'width' ];
 		$props        = $this->match_element_attributes( $html, 'iframe', $iframe_props );
@@ -203,7 +204,7 @@
 		}
 
 		$url        = $node->getAttribute( 'src' );
-		$attributes = $this->prepare_attributes( $url );
+		$attributes = $this->prepare_attributes( $url, $video_id );
 
 		if ( ! empty( $node->getAttribute( 'title' ) ) ) {
 			$attributes['title'] = $node->getAttribute( 'title' );
@@ -253,17 +254,12 @@
 	/**
 	 * Prepare attributes for amp-youtube component.
 	 *
-	 * @param string $url YouTube video URL.
+	 * @param string $url      YouTube video URL.
+	 * @param string $video_id YouTube video ID.
 	 *
 	 * @return array prepared arguments for amp-youtube component.
 	 */
-	private function prepare_attributes( $url ) {
-
-		$video_id = $this->get_video_id_from_url( $url );
-
-		if ( ! $video_id ) {
-			return [];
-		}
+	private function prepare_attributes( $url, $video_id ) {
 
 		$attributes = [
 			'data-videoid' => $video_id,
@@ -510,6 +506,6 @@
 			return $html;
 		}
 
-		return $this->render( $html, $attr['src'] );
+		return $this->render( $html, $attr['src'], $video_id );
 	}
 }
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -82,117 +94,313 @@ public function unregister_embed() {
-		if ( ! empty( $args['title'] ) ) {
-			$attributes['title'] = $args['title'];
+
+		if ( ! empty( $amp_node ) && is_a( $amp_node, 'DOMElement' ) ) {
```
Best to specify the full namespace of the class as to avoid any issues if this class were to be `namespace`d.

```suggestion
		if ( ! empty( $amp_node ) && is_a( $amp_node, '\DOMElement' ) ) {
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -82,117 +94,313 @@ public function unregister_embed() {
+
+		$attributes = $this->prepare_attributes( $url );
+
+		$iframe_props = [ 'title', 'height', 'width' ];
+		$props        = $this->match_element_attributes( $html, 'iframe', $iframe_props );
+		foreach ( $iframe_props as $iframe_prop ) {
+			if ( ! empty( $props[ $iframe_prop ] ) ) {
+				$attributes[ $iframe_prop ] = $props[ $iframe_prop ];
+			}
 		}
```
This could be moved to the `prepare_attributes()` method as it seems more relevant there.

dhaval-parekh replied:
These three attributes parse differently in `render()` (from HTML markup) and in `get_amp_component()` (from DOMElement). That is a reason it's parse saperatlly.

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -82,117 +94,313 @@ public function unregister_embed() {
 		}
 
-		$this->did_convert_elements = true;
+		$video_title = ( ! empty( $attributes['title'] ) ) ? $attributes['title'] : null;
```
Seems to be unused.

```suggestion
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -82,117 +94,313 @@ public function unregister_embed() {
+	}
+
+	/**
+	 * Placeholder component for amp YouTube component..
```
```suggestion
	 * Placeholder component for AMP YouTube component.
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -236,6 +444,51 @@ private function get_video_id_from_url( $url ) {
+		}
+
+		if ( ! empty( $parsed_url['fragment'] ) ) {
+			$regex = '/^t=(?<minutes>\d+)m(?<seconds>\d+)s$/';
```
Actually the `t` fragment turns out to be more complex than expected. All of these are valid:

URL | Time
---|---
https://www.youtube.com/watch?v=kfVsfOSbJY0#t=62   | 62 secs
https://www.youtube.com/watch?v=kfVsfOSbJY0#t=1m2s | 1 min 2 secs
https://www.youtube.com/watch?v=kfVsfOSbJY0#t=62s  | 62 secs
https://www.youtube.com/watch?v=kfVsfOSbJY0#t=1m   | 1 min
https://www.youtube.com/watch?v=kfVsfOSbJY0#t=1m2  | 1 mins 2 secs

So that complicates the regex a bit. This should do:

```suggestion
			$regex = '/^t=(?:(?<minutes>\d+)m)?(?:(?<seconds>\d+)s?)?$/';
```

dhaval-parekh committed:
Update prepare_attribute()

dhaval-parekh committed:
Update regex for to find seek time from URL

dhaval-parekh committed:
Add additional unit test cases for video seek time

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -82,117 +94,309 @@ public function unregister_embed() {
+				continue;
+			}
+
+			$attributes[ "data-param-$key" ] = esc_attr( $value );
```
```suggestion
			$attributes[ sanitize_key( "data-param-$key" ) ] = esc_attr( $value );
```

pierlon commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -82,117 +94,309 @@ public function unregister_embed() {
+			$attributes
+		);
+
+		if ( ! empty( $amp_node ) && is_a( $amp_node, '\DOMElement' ) ) {
```
```suggestion
		if ( ! empty( $amp_node ) && $amp_node instanceof Element ) {
```

pierlon reviewed (approved):
Thanks for your hard work on this @dhaval-parekh, passing this on to @westonruter for final review.

westonruter commented on includes/embeds/class-amp-youtube-embed-handler.php:
```diff
@@ -82,117 +94,311 @@ public function unregister_embed() {
+				continue;
+			}
+
+			$attributes[ sanitize_key( "data-param-$key" ) ] = esc_attr( $value );
```
Escaping shouldn't be done here because the returned attributes array may be sent into the DOM directly, in which case the escaping will be duplicated.

```suggestion
			$attributes[ sanitize_key( "data-param-$key" ) ] = $value;
```

westonruter reviewed (commented):
@dhaval-parekh A few tests are failing. It seems just the expected link in the text data needs to be updated.

dhaval-parekh committed:
Sanitize attribute name
Use original YouTube URL as placeholder when sanitizing raw embed
Fix phpcs alignment warning
Use constant instead of member variable
Let Document::fromNode() obtain the ownerDocument
Deduplicate iframe_props into constant

dhaval-parekh committed:
update unit test cases

westonruter committed:
Revert phpstan ignores

westonruter commented:
@dhaval-parekh The phpstan ignores that you added caused errors for me locally in the same way as they caused build failures on GHA. So I reverted them in 3ecad1e3fa9b44e2c7a0b66918099c96f60126d6. It may be that you have a stale copy of phpstan.  I have v0.12.76 (humm, which is also stale). I thought that `composer clear-cache && composer install` would get the latest version, but it didn't work for me. So you can do this instead: `rm vendor/bin/phpstan && composer install`.

westonruter commented:
Er, when I upgrade to `0.12.98` then I get errors:

```
 ------ -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   includes/embeds/class-amp-twitter-embed-handler.php
 ------ -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  24     PHPDoc type int|string of property AMP_Twitter_Embed_Handler::$DEFAULT_WIDTH is not covariant with PHPDoc type int of overridden property AMP_Base_Embed_Handler::$DEFAULT_WIDTH.
         💡 You can fix 3rd party PHPDoc types with stub files:
            https://phpstan.org/user-guide/stub-files
 ------ -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

 ------ ---------------------------------------------------
  Line   src/BackgroundTask/MonitorCssTransientCaching.php
 ------ ---------------------------------------------------
  112    No error to ignore is reported on line 112.
 ------ ---------------------------------------------------
```

So apparently GitHub's copy is stale and you _were_ on the latest version.

westonruter committed:
Try adding phpstan version to composer cache key

westonruter committed:
Revert "Revert phpstan ignores"

This reverts commit 7f21703d9563066e6fbc4fda7d070ef514fd6ca3.

westonruter committed:
Print phpstan version when running on GHA

westonruter committed:
Try removing phpstan before composer install

westonruter committed:
Fix yml syntax error

westonruter committed:
Try downloading a specific version of phpstan instead of latest

pierlon commented:
@westonruter As an alternative, might I suggest installing PHPStan from the `shivammathur/setup-php` GHA? Not providing a version for phpstan via that GHA should download the latest version. It should look something like:

```diff
diff --git a/.github/workflows/build-test-measure.yml b/.github/workflows/build-test-measure.yml
--- a/.github/workflows/build-test-measure.yml	(revision 08017ae8e1da05ae8e445e51807610e4e5365599)
+++ b/.github/workflows/build-test-measure.yml	(date 1631210205414)
@@ -214,6 +214,7 @@
           # phpstan requires PHP 7.1+.
           php-version: 7.4
           extensions: dom, iconv, json, libxml, zip
+          tools: phpstan
 
       - name: Get Composer Cache Directory
         id: composer-cache
@@ -231,7 +232,7 @@
         run: composer install
 
       - name: Static Analysis (PHPStan)
-        run: vendor/bin/phpstan analyse
+        run: phpstan analyse
 
 #-----------------------------------------------------------------------------------------------------------------------
 
```

westonruter committed:
Try installing phpstan via shivammathur/setup-php

Co-authored-by: Pierre Gordon <pierregordon@protonmail.com>

westonruter committed:
Fix issues identified by phpstan

westonruter committed:
Give more explicit array phpdoc type

westonruter committed:
Remove unreachable code

westonruter committed:
Opt to pass-through iframe for unrecognized YouTube URL rather than convert to link

westonruter committed:
Remove condition which cannot be reached since video ID is always provided

westonruter committed:
Use coversDefaultClass

westonruter committed:
Prevent generating closing IMG tags

westonruter committed:
Let placeholder link point to YouTube permalink and not iframe src

westonruter committed:
Fix adding start to placeholder link; improve coverage

westonruter committed:
Make placeholder dependency on video ID more explicit

westonruter commented:
💥  ![image](https://user-images.githubusercontent.com/134745/132770549-fbcc9c33-6e08-45f9-9c85-dfb57d634fc7.png)

westonruter reviewed (approved):
Whew! This was a deceptively complicated one!
